src/backend/distributed/planner/recursive_planning.c

```diff
@@ -252,10 +317,605 @@ RecursivelyPlanSubqueriesAndCTEs(Query *query, RecursivePlanningContext *context
 		RecursivelyPlanAllSubqueries((Node *) query->jointree->quals, context);
 	}
 
+	/*
+	 * If the query doesn't have distribution key equality,
+	 * recursively plan some of its subqueries.
+	 */
+	if (ShouldRecursivelyPlanNonColocatedSubqueries(query, context))
+	{
+		RecursivelyPlanNonColocatedSubqueries(query, context);
+	}
+
 	return NULL;
 }
 
 
+/*
+ * ShouldRecursivelyPlanNonColocatedSubqueries returns true if the input query contains joins
+ * that are not on the distribution key.
+ * *
+ * Note that at the point that this function is called, we've already recursively planned all
+ * the leaf subqueries. Thus, we're actually checking whether the joins among the subqueries
+ * on the distribution key or not.
+ */
+static bool
+ShouldRecursivelyPlanNonColocatedSubqueries(Query *subquery,
+											RecursivePlanningContext *context)
+{
+	/*
+	 * If the input query already contains the equality, simply return since it is not
+	 * possible to find any non colocated subqueries.
+	 */
+	if (context->allDistributionKeysInQueryAreEqual)
+	{
+		return false;
+	}
+
+	/*
+	 * This check helps us in two ways:
+	 *   (i) We're not targeting queries that don't include subqueries at all,
+	 *       they should go through regular planning.
+	 *  (ii) Lower level subqueries are already recursively planned, so we should
+	 *       only bother non-colocated subquery joins, which only happens when
+	 *       there are subqueries.
+	 */
+	if (!ContainsSubquery(subquery))
+	{
+		return false;
+	}
+
+	/* direct joins with local tables are not supported by any of Citus planners */
+	if (FindNodeCheckInRangeTableList(subquery->rtable, IsLocalTableRTE))
+	{
+		return false;
+	}
+
+	/*
+	 * Finally, check whether this subquery contains distribution key equality or not.
+	 */
+	if (!AllDistributionKeysInSubqueryAreEqual(subquery,
+											   context->plannerRestrictionContext))
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * ContainsSubquery returns true if the input query contains any subqueries
+ * in the FROM or WHERE clauses.
+ */
+static bool
+ContainsSubquery(Query *query)
+{
+	return ContainsSubqueryInFromClause(query) || ContainsSubqueryInWhereClause(query);
+}
+
+
+/*
+ * ContainsSubqueryInFromClause returns true if the input query contains any subqueries
+ * in the FROM clause.
+ */
+static bool
+ContainsSubqueryInFromClause(Query *query)
+{
+	List *rangeTableList = query->rtable;
+	List *joinTreeTableIndexList = NIL;
+	ListCell *joinTreeTableIndexCell = NULL;
+
+	/*
+	 * Extract all range table indexes from the join tree. Note that here we
+	 * only walk over range table entries at this level and do not recurse into
+	 * subqueries.
+	 */
+	ExtractRangeTableIndexWalker((Node *) query->jointree, &joinTreeTableIndexList);
```

Comment:
Now we're still building a list and it's a lot of (duplicate) code. Why not something like:

```
bool
JoinTreeContainsSubquery(Node *joinTreeNode, Query *query)
{
    if (joinTreeNode == NULL)
    {
        return false;
    }

    if (IsA(joinTreeNode, RangeTblRef))
    {
        RangeTblRef *rangeTableRef = (RangeTblRef *) joinTreeNode;
        RangeTblEntry *rangeTableEntry = rt_fetch(rangeTableRef->rtindex, query->rtable);
        if (rangeTableEntry->rtekind == RTE_SUBQUERY)
        {
            return true;
        }
    }

    return expression_tree_walker(joinTreeNode, JoinTreeContainsSubqueryWalker, query);
}
```

Reply:
> Now we're still building a list and it's a lot of (duplicate) code. 

Oops, I copy & pasted the code and ignored the list